llvm/test/CodeGen/NVPTX/reduction-intrinsics.ll

```diff
@@ -0,0 +1,1908 @@
+define float @reduce_fmax_float(<8 x float> %in) {
+; CHECK-SM80-LABEL: reduce_fmax_float(
+; CHECK-SM80:       {
+; CHECK-SM80-NEXT:    .reg .f32 %f<16>;
+; CHECK-SM80-EMPTY:
+; CHECK-SM80-NEXT:  // %bb.0:
+; CHECK-SM80-NEXT:    ld.param.v4.f32 {%f5, %f6, %f7, %f8}, [reduce_fmax_float_param_0+16];
+; CHECK-SM80-NEXT:    ld.param.v4.f32 {%f1, %f2, %f3, %f4}, [reduce_fmax_float_param_0];
+; CHECK-SM80-NEXT:    max.f32 %f9, %f1, %f2;
+; CHECK-SM80-NEXT:    max.f32 %f10, %f9, %f3;
+; CHECK-SM80-NEXT:    max.f32 %f11, %f10, %f4;
+; CHECK-SM80-NEXT:    max.f32 %f12, %f11, %f5;
+; CHECK-SM80-NEXT:    max.f32 %f13, %f12, %f6;
+; CHECK-SM80-NEXT:    max.f32 %f14, %f13, %f7;
+; CHECK-SM80-NEXT:    max.f32 %f15, %f14, %f8;
+; CHECK-SM80-NEXT:    st.param.f32 [func_retval0], %f15;
+; CHECK-SM80-NEXT:    ret;
+;
+; CHECK-SM100-LABEL: reduce_fmax_float(
+; CHECK-SM100:       {
+; CHECK-SM100-NEXT:    .reg .f32 %f<13>;
+; CHECK-SM100-EMPTY:
+; CHECK-SM100-NEXT:  // %bb.0:
+; CHECK-SM100-NEXT:    ld.param.v4.f32 {%f5, %f6, %f7, %f8}, [reduce_fmax_float_param_0+16];
+; CHECK-SM100-NEXT:    ld.param.v4.f32 {%f1, %f2, %f3, %f4}, [reduce_fmax_float_param_0];
+; CHECK-SM100-NEXT:    max.f32 %f9, %f1, %f2, %f3;
+; CHECK-SM100-NEXT:    max.f32 %f10, %f9, %f4, %f5;
+; CHECK-SM100-NEXT:    max.f32 %f11, %f10, %f6, %f7;
+; CHECK-SM100-NEXT:    max.f32 %f12, %f11, %f8;
+; CHECK-SM100-NEXT:    st.param.f32 [func_retval0], %f12;
+; CHECK-SM100-NEXT:    ret;
+  %res = call float @llvm.vector.reduce.fmax(<8 x float> %in)
+  ret float %res
+}
+
+; Check tree reduction.
+define float @reduce_fmax_float_reassoc(<8 x float> %in) {
+; CHECK-SM80-LABEL: reduce_fmax_float_reassoc(
+; CHECK-SM80:       {
+; CHECK-SM80-NEXT:    .reg .f32 %f<16>;
+; CHECK-SM80-EMPTY:
+; CHECK-SM80-NEXT:  // %bb.0:
+; CHECK-SM80-NEXT:    ld.param.v4.f32 {%f5, %f6, %f7, %f8}, [reduce_fmax_float_reassoc_param_0+16];
+; CHECK-SM80-NEXT:    ld.param.v4.f32 {%f1, %f2, %f3, %f4}, [reduce_fmax_float_reassoc_param_0];
+; CHECK-SM80-NEXT:    max.f32 %f9, %f7, %f8;
+; CHECK-SM80-NEXT:    max.f32 %f10, %f5, %f6;
+; CHECK-SM80-NEXT:    max.f32 %f11, %f10, %f9;
+; CHECK-SM80-NEXT:    max.f32 %f12, %f3, %f4;
+; CHECK-SM80-NEXT:    max.f32 %f13, %f1, %f2;
+; CHECK-SM80-NEXT:    max.f32 %f14, %f13, %f12;
+; CHECK-SM80-NEXT:    max.f32 %f15, %f14, %f11;
+; CHECK-SM80-NEXT:    st.param.f32 [func_retval0], %f15;
+; CHECK-SM80-NEXT:    ret;
+;
+; CHECK-SM100-LABEL: reduce_fmax_float_reassoc(
+; CHECK-SM100:       {
+; CHECK-SM100-NEXT:    .reg .f32 %f<13>;
+; CHECK-SM100-EMPTY:
+; CHECK-SM100-NEXT:  // %bb.0:
+; CHECK-SM100-NEXT:    ld.param.v4.f32 {%f5, %f6, %f7, %f8}, [reduce_fmax_float_reassoc_param_0+16];
+; CHECK-SM100-NEXT:    ld.param.v4.f32 {%f1, %f2, %f3, %f4}, [reduce_fmax_float_reassoc_param_0];
+; CHECK-SM100-NEXT:    max.f32 %f9, %f4, %f5, %f6;
+; CHECK-SM100-NEXT:    max.f32 %f10, %f1, %f2, %f3;
+; CHECK-SM100-NEXT:    max.f32 %f11, %f10, %f9, %f7;
+; CHECK-SM100-NEXT:    max.f32 %f12, %f11, %f8;
```

Comment:
This is a good example of tree reduction being split across multiple loads that may arrive at different times:

```
max2(
  max3(
    max3(f4,f5,f6),
    max3(f1,f2,f3).
    f7),
  f8)
```

I wonder if there would be an observable performance difference vs:
```
max3(
  max3(f1,f2,f3),
  max3(f5,f6,f7)
  max2(f4, f8)
)
```
It would potentially have one instruction shorted data dependency chain.